Failed to generate tls-crypt-v2 client key

[INFO] Generating client certificate...
> ./easyrsa --batch build-client-full SPB nopass
[OK] Client SPB added and is valid for 3650 days.
> cp /etc/openvpn/server/client-template.txt /root/SPB.ovpn
[ERROR] Failed to generate tls-crypt-v2 client key
        Check the log file for details: openvpn-install.log

OS Ubuntu 25.04  and OS Ubuntu 25.10 
2025-12-13 23:30:41 Cannot open file '/tmp/tmp.qGRtkdczID' for write: Permission denied (errno=13)
2025-12-13 23:30:41 ERROR: could not write client key file
2025-12-13 23:30:41 Exiting due to fatal error

--- a/.github/workflows/docker-test.yml
+++ b/.github/workflows/docker-test.yml
@@ -30,6 +30,8 @@ jobs:
             image: ubuntu:22.04
           - name: ubuntu-24.04
             image: ubuntu:24.04
+          - name: ubuntu-25.10
+            image: ubuntu:25.10
           - name: debian-11
             image: debian:11
           - name: debian-12

--- a/openvpn-install.sh
+++ b/openvpn-install.sh
@@ -1140,9 +1140,10 @@ function installOpenVPN() {
 		esac
 
 		# Generate a random, alphanumeric identifier of 16 characters for CN and one for server name
-		SERVER_CN="cn_$(head /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 16 | head -n 1)"
+		# Note: 2>/dev/null suppresses "Broken pipe" errors from fold when head exits early
+		SERVER_CN="cn_$(head /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 16 2>/dev/null | head -n 1)"
 		echo "$SERVER_CN" >SERVER_CN_GENERATED
-		SERVER_NAME="server_$(head /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 16 | head -n 1)"
+		SERVER_NAME="server_$(head /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 16 2>/dev/null | head -n 1)"
 		echo "$SERVER_NAME" >SERVER_NAME_GENERATED
 
 		# Create the PKI, set up the CA, the DH params and the server certificate
@@ -1707,8 +1708,13 @@ function generateClientConfig() {
 
 		case $tls_sig in
 		1)
-			# Generate per-client tls-crypt-v2 key using secure temp file
-			tls_crypt_v2_tmpfile=$(mktemp)
+			# Generate per-client tls-crypt-v2 key in /etc/openvpn/server/
+			# Using /tmp would fail on Ubuntu 25.04+ due to AppArmor restrictions
+			tls_crypt_v2_tmpfile=$(mktemp /etc/openvpn/server/tls-crypt-v2-client.XXXXXX)
+			if [[ -z "$tls_crypt_v2_tmpfile" ]] || [[ ! -f "$tls_crypt_v2_tmpfile" ]]; then
+				log_error "Failed to create temporary file for tls-crypt-v2 client key"
+				exit 1
+			fi
 			if ! openvpn --tls-crypt-v2 /etc/openvpn/server/tls-crypt-v2.key \
 				--genkey tls-crypt-v2-client "$tls_crypt_v2_tmpfile"; then
 				rm -f "$tls_crypt_v2_tmpfile"

